### PR TITLE
Improve the memory usage and performance of the cache plugin

### PR DESCRIPTION
**Pull request by charlievieth**
### 1. Why is this pull request needed and what does it do?
This PR reduces the memory usage of the cache plugin (14 vs 8 allocs and 540 vs 327 B) and improves performance as a side effect.

**PR:**
```
goos: darwin
goarch: amd64
pkg: github.com/coredns/coredns/plugin/cache
BenchmarkCacheResponse/NoPrefetch-12         	 1601374	       732 ns/op	     327 B/op	       8 allocs/op
BenchmarkCacheResponse/Prefetch-12           	 1434120	       830 ns/op	     327 B/op	       8 allocs/op
BenchmarkHash-12                             	130255627	         9.25 ns/op	       0 B/op	       0 allocs/op
PASS
```

master@ 004c5fca9d7dfb8b58d608e4d50833b40953441f
```
goos: darwin
goarch: amd64
pkg: github.com/coredns/coredns/plugin/cache
BenchmarkCacheResponse-12    	 1181955	       971 ns/op	     540 B/op	      14 allocs/op
PASS
```

### 2. Which issues (if any) are related?
N/A
### 3. Which documentation changes (if any) need to be made?
N/A
### 4. Does this introduce a backward incompatible change or deprecation?
N/A

### CODE REVIEW

**Comment by miekg** [comment]
Note there is also this closed PR https://github.com/coredns/coredns/pull/2527/files that tackles this on a more fundamental level, but not storing parts of a message that don't need to be stored. That didn't get merged because getting actual memory improvement data was hard.

File: plugin/cache/cache.go [hunk 1]
```diff
@@ -76,21 +75,39 @@ func key(qname string, m *dns.Msg, t response.Type, do bool) (bool, uint64) {
 	return true, hash(qname, m.Question[0].Qtype, do)
 }
 
-var one = []byte("1")
-var zero = []byte("0")
+type fnv64 uint64
```
Comment (miekg) [inline plugin/cache/cache.go:78]: This needs some documentation on what's being done here. Is there any (semi) official hash algo that we can use? And why isn't fnv in the std library more efficient?

File: plugin/cache/handler.go [hunk 2]
```diff
@@ -29,28 +29,10 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
...
-
-					// When prefetching we loose the item i, and with it the frequency
-					// that we've gathered sofar. See we copy the frequencies info back
-					// into the new item that was stored in the cache.
-					if i1 := c.exists(state); i1 != nil {
-						i1.Freq.Reset(now, i.Freq.Hits())
-					}
-				}(cw)
-			}
+		if c.shouldPrefetch(i, now) {
```
Comment (miekg) [inline plugin/cache/handler.go:34]: could change, but unrelated. Can you pull it into a new PR?

File: plugin/cache/item.go [hunk 3]
```diff
@@ -9,15 +9,13 @@ import (
...
-	AuthenticatedData  bool
-	RecursionAvailable bool
 	Answer             []dns.RR
 	Ns                 []dns.RR
 	Extra              []dns.RR
-
-	origTTL uint32
-	stored  time.Time
-
+	AuthenticatedData  bool
```
Comment (miekg) [inline plugin/cache/item.go:15]: unrelated change and these where grouped with the other msg related fields (i.e. there was some significance in that)

File: plugin/cache/item.go [hunk 4]
```diff
@@ -62,9 +60,14 @@ func (i *item) toMsg(m *dns.Msg, now time.Time) *dns.Msg {
 	m1.RecursionAvailable = i.RecursionAvailable
 	m1.Rcode = i.Rcode
 
-	m1.Answer = make([]dns.RR, len(i.Answer))
-	m1.Ns = make([]dns.RR, len(i.Ns))
-	m1.Extra = make([]dns.RR, len(i.Extra))
+	m1.Answer = i.Answer
+	m1.Ns = i.Ns
+	m1.Extra = i.Extra
```
Comment (rdrozhdzh) [inline plugin/cache/item.go:65]: do you really need these 3 lines above?

File: plugin/cache/item.go [hunk 5]
```diff
@@ -62,9 +60,14 @@ func (i *item) toMsg(m *dns.Msg, now time.Time) *dns.Msg {
...
+
+	a := make([]dns.RR, len(i.Answer)+len(i.Ns)+len(i.Extra))
+	m1.Answer, a = a[:len(i.Answer)], a[len(i.Answer):]
```
Comment (rdrozhdzh) [inline plugin/cache/item.go:68]: This is not quite safe. If any code later will want to append an RR to answer, it will overwrite NS record.
Consider assigning slice with capacity specified like
```
m1.Answer = a[:len(i.Answer):len(i.Answer)]
```

File: plugin/cache/cache.go (hunk 1 above)
Comment (charlievieth) [reply plugin/cache/cache.go:78]: I'll document it, this is the same `fnv` function as the stdlib, but has a WriteByte and WriteString method.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Improve the m..." ]
>charlievieth commented on this pull request.
>
>
>
>> @@ -76,21 +75,39 @@ func key(qname string, m *dns.Msg, t response.Type, do bool) (bool, uint64) {
> 	return true, hash(qname, m.Question[0].Qtype, do)
> }
>
>-var one = []byte("1")
>-var zero = []byte("0")
>+type fnv64 uint64
>
>I'll document it, this is the same `fnv` function as the stdlib, but has a WriteByte and WriteString method.

I'm not a fan of rewriting parts of the stdlib, we'll now miss out any fixes 
(like ASM) they apply to this. Is there a smarter way of doing this? Dropping 
FNV all together?

**Comment by miekg** [comment]
I'm moving some of this into their own PRs. For one is prefetch closure, see: #3353

**Comment by miekg** [comment]
closing what's left of this pr
